hub/src/server/server.ts

```diff
@@ -240,7 +256,45 @@ export class HubServer {
       }
     }
 
-    const readURL = await this.driver.performWrite(writeCommand)
+    // Use the stream pipe API to monitor a stream with correct backpressure handling. This 
+    // avoids buffering entire streams in memory and hooks up all the correct events for 
+    // cleanup and error handling. See https://nodejs.org/api/stream.html#stream_three_states
+
+    // Two stream listeners are required to enforce size limits: the original stream reader 
+    // implemented by drivers for uploading to storage, and another for monitoring upload size. 
+    // Stream listeners are greedy so two PassThrough streams are used, and the pipeline API
+    // is setup *before* reading so that both listeners receive all notifications. 
+    // See https://stackoverflow.com/a/51143558/794962
+
+    // Create a PassThrough stream to monitor streaming size. 
+    const monitorStream = new PassThrough()
+    const monitorPipeline = pipelineAsync(stream, monitorStream)
+
+    // Create a PassThrough stream to give to driver for uploading to storage backend. 
+    const uploadStream = new PassThrough()
+    const uploadPipeline = pipelineAsync(stream, uploadStream)
+
+    // Now that the PassThrough streams and pipes are setup, the data event listener
+    // for the size monitor can be setup without either stream listeners missing any chunks. 
+    let monitoredContentSize = 0
+    monitorStream.on('data', (chunk: Buffer) => {
```

Comment:
Reviewer note:
This creates two reader streams for a file upload. This has the potential, depending on node/V8 optimizations, to briefly duplicate network chunk streams in memory. But in general, this is the recommended approach to properly handling backpressure and monitoring stream progress when there are multiple consumers.